Runtime/LLMCharacter.cs

```diff
@@ -490,44 +453,33 @@ protected string SlotContent(SlotResult result)
         /// <returns>the LLM response</returns>
         public async Task<string> Chat(string query, Callback<string> callback = null, EmptyCallback completionCallback = null, bool addToHistory = true)
         {
-            // handle a chat message by the user
-            // call the callback function while the answer is received
-            // call the completionCallback function when the answer is fully received
             await LoadTemplate();
             if (!CheckTemplate()) return null;
             if (!await InitNKeep()) return null;
+           
+            var playerMessage = new ChatMessage() { role = playerRole, content = query };
 
-            string json;
-            await chatLock.WaitAsync();
-            try
-            {
-                AddPlayerMessage(query);
-                string prompt = template.ComputePrompt(chat, playerName, AIName);
-                json = JsonUtility.ToJson(GenerateRequest(prompt));
-                chat.RemoveAt(chat.Count - 1);
-            }
-            finally
-            {
-                chatLock.Release();
-            }
+            // Setup the full list of messages for the current request
+            List<ChatMessage> promptMessages = chatHistory ? chatHistory.GetChatMessages() : new List<ChatMessage>();
+            promptMessages.Insert(0, GetSystemPromptMessage());
+            promptMessages.Add(playerMessage);
 
-            string result = await CompletionRequest(json, callback);
+            // Prepare the request
+            string formattedPrompt = template.ComputePrompt(promptMessages, playerRole, aiRole);
```

Comment:
We can change the ComputePrompt to take as input the system prompt and the player message as additional parameters instead of creating the promptMessages.
I can take that in a separate PR

Reply:
Shouldn't we keep the ComputePrompt more generic? All it needs is a list of messages, whoever is calling it can worry about what order the messages are in.